WIP: add more tests

sobuch opened the pull request.

sobuch committed:
fix in entropy init + some tests

sobuch commented:
discovered problem in entropy init, but it can wait here with other possible minor issues tests could uncover until the merge. Its not in any way critical.

sobuch commented:
If you want more tests for some reason, feel free to use this branch. I will continuously update it to see whether test coverage increases.

sobuch commented:
Also, @mvondracek as we discussed earlier, that it would be useful to run cli tests twice to record coverage - does this need some support on the script side or is it something that can be configured here in github?

mvondracek commented on PA193_mnemonic_Slytherin/test_mnemonic.py:
```diff
@@ -227,39 +229,67 @@ def test_verify_invalid_mnemonic_too_long(self):
+            try:
+                Entropy(unhexlify(test_vector[0]))
+            except ValueError:
+                self.fail("Instantiation failed unexpectedly!")
```
If the test expects line instantiation to succeed, try catch it? If there's error and uncaught exception is raised, the test fails, which is what we want.

mvondracek reviewed (commented):
Note: please write commits so that it brings one small change for each. For example e69e6eb98059061f04c2edc44b0816ff072185a8 again fixes something in code and then adds some other tests. We recently had to deal with reverting commit which included more than one thing...

mvondracek commented:
> 
> 
> Also, @mvondracek as we discussed earlier, that it would be useful to run cli tests twice to record coverage - does this need some support on the script side or is it something that can be configured here in github?

I'll set it up.

mvondracek committed:
fix test_verify_invalid_mnemonic_too_long

OverflowError
fix: 339b9c1f67537679858f0ea8a03e5737d830a941
related: #26, #28

mvondracek committed:
cs docstrings

mvondracek committed:
TestMnemonic: test naming

mvondracek committed:
TestMnemonic: removed unnecessary try-except

mvondracek committed:
TestMnemonic: removed unnecessary try-except

mvondracek commented:
@sobuch https://github.com/mvondracek/PA193_mnemonic_Slytherin/commit/bc8ccc731a03c4b61b4ded109c1b9faf062a2cb5#commitcomment-35625869

mvondracek committed:
Merge branch 'task-40-more-tests' of https://github.com/mvondracek/PA193_mnemonic_Slytherin into task-40-more-tests

mvondracek committed:
fix test_secure_seed_compare
Moved from TestMnemonicInternal.test_secure_seed_compare to TestSeed.test___eq__.
Removed unused code which is incompatible with class model (#36).

mvondracek committed:
TestSeed.test___eq__ removed unused code which is already tested in TestSeed.test___init___invalid_argument

mvondracek committed:
TestSeed: add setUp

mvondracek committed:
TestSeed: add test__ne__

mvondracek committed:
TestSeed: fix naming of tests

mvondracek committed:
cleanup

mvondracek committed:
TestSeed: ignore static type check where TypeError is expected

mvondracek committed:
Mnemonic: error messages in raised exceptions

mvondracek committed:
Entropy.checksum: removed unnecessary length parameter

mvondracek committed:
TestMnemonic: test___init__: add whitespace combinations

mvondracek committed:
TestMnemonic: add test___init___invalid_argument

mvondracek committed:
TestEntropy: add test_checksum, add Mnemonic.checksum

mvondracek committed:
TestEntropy: test_toMnemonic

fixes: e69e6eb98059061f04c2edc44b0816ff072185a8

mvondracek committed:
TestEntropy: fix naming of tests

mvondracek committed:
TestEntropy: ignore static type check where TypeError is expected

mvondracek committed:
Entropy.__init__: simplified condition on entroly length

mvondracek committed:
Merge remote-tracking branch 'refs/remotes/origin/task-40-more-tests' into refactor/mnemonic

mvondracek committed:
cleanup

mvondracek committed:
TestMnemonic.test___init___invalid_argument: update expected error based on b40219dba5550bad01b485502f214922928b64f1

mvondracek committed:
TestMnemonic.test___init___invalid_argument: test for specific errors

mvondracek committed:
TestMnemonic: add test_checksum

TODO Could we check `Mnemonic.checksum` without `Entropy.checksum`? See
`TestEntropy.test_checksum`.

mvondracek committed:
TestMnemonic: add test_toEntropy

fixes: e69e6eb98059061f04c2edc44b0816ff072185a8

mvondracek committed:
TestMnemonic: add test_toSeed

fixes: e69e6eb98059061f04c2edc44b0816ff072185a8

mvondracek committed:
cleanup

mvondracek committed:
Mnemonic.toSeed: check type of `seed_password` argument

mvondracek committed:
cleanup

mvondracek committed:
Mnemonic.toEntropy + Entropy.toMnemonic: return deep copy

mvondracek committed:
remove `self` reassignment in `__init__` without effect

mvondracek committed:
fix: call init of super class

mvondracek committed:
fix: class and method names according to PEP8

mvondracek committed:
renamec `DictionaryAccess` to `_DictionaryAccess`

mvondracek committed:
fix: typing compatible with Python 3.5

mvondracek committed:
Merge pull request #43 from mvondracek/refactor/mnemonic

Refactor/mnemonic

mvondracek committed:
TestSeed: add tests for not an instance of Seed

mvondracek committed:
cleanup

mvondracek committed:
cleanup

mvondracek committed:
text_mnemoniccli: run program as subprocess and also inside tests to count test coverage

mvondracek committed:
cleanup

mvondracek committed:
fix: rename `assert_program_inside_tests` to `assert_program_entry_point`

Method `assert_program_inside_tests` was not a separate test, but just
method used by other tests. It was incorrectly included in tests,
because of `test` in its name.

mvondracek commented:
This is still work in progress.

mvondracek committed:
TestPublicFunctions: add test_generate_invalid_arguments

mvondracek committed:
TestPublicFunctions: add test_recover_invalid_arguments

mvondracek committed:
TestPublicFunctions: add test_verify_invalid_arguments

mvondracek committed:
cleanup

mvondracek committed:
TestMnemonic: add test_checksum_invalid_mnemonic

mvondracek committed:
tests: add Test_DictionaryAccess

mvondracek committed:
tests: _DictionaryAccess: test invalid_dictionary_lines

mvondracek committed:
fix: mnemoniccli: -g and -r require all -e -m -s

sobuch commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -187,6 +187,49 @@ def __init__(self, mnemonic: str):
                              .format(checksum_included, checksum_computed))
         self = mnemonic
 
+    @staticmethod
```
why should there be this as a static method? every mnemonic necesarily computes its entropy upon instantiation, to verify validity. Moreover 'checksum' returns the checksum of Entropy, it is not checksum of the mnemonic.

mvondracek replied:
This function is not dependent on any instance of Mnemonic. It takes mnemonic phrase as string and computes checksum based on words from the mnemonic phrase and given dictionary. It is used to test `Entropy.checksum()`, because `Entropy.checksum()` computes checksum using hashing bytes of entropy. Based on our testing data, we know which mnemonic phrase corresponds to which entropy bytes, therefore we can test `Entropy.checksum()`, but we need a way how to get checksum based on mnemonic phrase **without generating entropy and hashing entropy bytes to get checksum**.

sobuch replied:
but you are talking about tests, we can test entropy.checksum without this, we just need pairs of checksum - entropy, same as we dont have two functions to generate seed from mnemonic and still can test it. I dont think this method provides any functionality just duplicates existing one and existing code

mvondracek replied:
> we can test entropy.checksum without this, we just need pairs of checksum - entropy

But we don't have those pairs, that's the reason for this method. 1) We can extract checksum from *mnemonic phrase*. 2) We can calculate checksum from `Entropy`.

> we dont have two functions to generate seed from mnemonic and still can test it

That's different, because we already have testing data with related `Mnemonic`, `Entropy`, and `Seed`. 

> duplicates existing one and existing code

This points out, that access to dictionary hidden inside `__init__` is not a good design, because the dictionary cannot be reused and cannot be reached from different contexts than `__init__`.

How do you suggest we would test `Entropy.checksum()` without this method?

sobuch replied:
you can use such a function to generate the test data, and then test it the same way as public methods. But I dont think it should be present in the library. Maybe it could be part of the tests, but even then it wouldnt make sense to me as, it generates the checksum the same way, just in different function.

> This points out, that access to dictionary hidden inside __init__ is not a good design, because the dictionary cannot be reused and cannot be reached from different contexts than __init__.

It might not be a good design for other reasons, but I dont think this is the case, you could duplicate any object code this way.

mvondracek replied:
> it generates the checksum the same way, just in different function

It does not, that's the point. It *"computes checksum based on words from the mnemonic phrase and given dictionary"*. On the other hand *"`Entropy.checksum()` computes checksum using hashing bytes of entropy"*. https://github.com/mvondracek/PA193_mnemonic_Slytherin/pull/41#discussion_r339294934

I can move it out from `mnemonic` to tests so it's not part of `Mnemonic` interface.

sobuch replied:
> It does not, that's the point. It "computes checksum based on words from the mnemonic phrase and given dictionary". On the other hand "Entropy.checksum() computes checksum using hashing bytes of entropy". #41 (comment)

how is it not the same code? you literally marked it as '# region TODO copied from _DictionaryAccess.__init__' and '# region TODO copied from Mnemonic.__init__'. I mean if you called Mnemonic.init and Mnemonic.checksum with the same string as argument, it would do the same things, no? I am not saying it generates checksum same as Entropy.checksum, but it does generate it the exact same way as it is done in Mnemonic.init

mvondracek replied:
`Mnemonic.__init__` and `Mnemonic.checksum` do internally similar things. `Mnemonic.checksum` gets checksum in different way than `Entropy.checksum`. Before this method, **you could not extract checksum from `Mnemonic`**. Benefit of this method is that we could test `Entropy.checksum`. Anyway, I just moved the code to tests

mvondracek replied:
Let's me know if this is solved from your view and whether this PR could be merged or if you have any other feedback.

mvondracek committed:
fix: test_mnemoniccli: -g and -r require all -e -m -s

mvondracek committed:
test_mnemoniccli: test_arguments_error: add more invalid argument combinations

mvondracek committed:
Merge branch 'dev' into task-40-more-tests

mvondracek committed:
test_mnemoniccli: add tests for generate, recover, verify

mvondracek committed:
fix: test names PEP8

mvondracek committed:
Merge branch 'dev' into task-40-more-tests

mvondracek committed:
tests: use single default valid testing data

mvondracek committed:
test_mnemonic: noinspection SpellCheckingInspection of testing data

mvondracek committed:
cleanup

mvondracek committed:
`catch StopIteration` from bed643598d635f3ddd48d1aafb82b73192641712

mvondracek commented:
Closes #40.

mvondracek committed:
Merge pull request #54 from mvondracek/bug-44

tests: _DictionaryAccess: test invalid_dictionary_lines

mvondracek committed:
test_mnemonic: fix test names PEP8

mvondracek committed:
fix: reference to VALID_MNEMONIC_PHRASE_TREZOR in TestMnemonic.test_checksum_invalid_dictionary_lines

mvondracek committed:
TestMain.test_arguments_error: add test for too long password

mvondracek committed:
test_mnemoniccli: add tests of actions with invalid input data

test_generate_invalid_entropy, test_recover_invalid_mnemonic,
test_verify_invalid_mnemonic, test_verify_invalid_seed

mvondracek committed:
test_mnemoniccli.TestMain: add test_verify_seeds_do_not_match

mvondracek committed:
test_mnemoniccli.TestMain: add test_verify_missing_seed_file, test_verify_missing_mnemonic_file

mvondracek committed:
Config.Format: read_mode, write_mode as property

mvondracek committed:
cleanup

mvondracek commented:
Tests, fixes of bugs, relevant refactoring of other code, and **99 % test coverage**. 🎉 
Some tests could be refactored for better structure, but I would merge it now in order to start with fixing #50, #51, #52, #55, ...

@sobuch, as you created this PR, you cannot set approval via GitHub UI, so please leave your feedback here in comments. :)

mvondracek commented:
Closes #40.

lsolodkova commented on PA193_mnemonic_Slytherin/test_mnemonic.py:
```diff
@@ -194,104 +170,415 @@ def test_secure_seed_compare(self):
+        """Too long mnemonic phrase."""
         with self.assertRaises(ValueError):
-            mnemonic.toSeed(password)
+            Mnemonic('a' * 1024 * 1024 * 1024 * 2)  # 2 GB
```
Failed with OverflowError, please, reduce;)

mvondracek replied:
Done. 0cf1235

mvondracek committed:
TestMnemonic.test___init___too_long_str: reduce size of mnemonic phrase to 1 MB

mvondracek committed:
Moved `mnemonic.Mnemonic.checksum` to `test_mnemonic.extract_checksum`

sobuch commented:
I think the tests look great. Lets merge this

mvondracek commented:
Waits for review from @lsolodkova
